processing/src/main/java/org/apache/druid/segment/join/JoinFilterAnalyzer.java

```diff
@@ -0,0 +1,739 @@
+             Objects.equals(getJoinTableFilter(), that.getJoinTableFilter()) &&
+             Objects.equals(getPushDownVirtualColumns(), that.getPushDownVirtualColumns());
+    }
+
+    @Override
+    public int hashCode()
+    {
+      return Objects.hash(getBaseTableFilter(), getJoinTableFilter(), getPushDownVirtualColumns());
+    }
+  }
+
+  /**
+   * Analyze a single filter clause from a filter that is in conjunctive normal form (AND of ORs),
+   * returning a JoinFilterAnalysis that contains a possible filter rewrite and information on how to handle the filter.
+   *
+   * @param adapter          Adapter for the join
+   * @param filterClause     Individual filter clause from a filter that is in CNF
+   * @param prefixes         Map of table prefixes
+   * @param equiconditions   Equicondition map
+   * @param correlationCache Cache of column correlation analyses
+   *
+   * @return a JoinFilterAnalysis that contains a possible filter rewrite and information on how to handle the filter.
+   */
+  private static JoinFilterAnalysis analyzeJoinFilterClause(
+      HashJoinSegmentStorageAdapter adapter,
+      Filter filterClause,
+      Map<String, JoinableClause> prefixes,
+      Map<String, Expr> equiconditions,
+      Map<String, List<JoinFilterColumnCorrelationAnalysis>> correlationCache
+  )
+  {
+    // NULL matching conditions are not currently pushed down
+    if (filterMatchesNull(filterClause)) {
+      return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+    }
+
+    // Currently we only support rewrites of selector filters and selector filters within OR filters.
+    if (filterClause instanceof SelectorFilter) {
+      return rewriteSelectorFilter(
+          adapter,
+          filterClause,
+          prefixes,
+          equiconditions,
+          correlationCache
+      );
+    }
+
+    if (filterClause instanceof OrFilter) {
+      return rewriteOrFilter(
+          adapter,
+          (OrFilter) filterClause,
+          prefixes,
+          equiconditions,
+          correlationCache
+      );
+    }
+
+    for (String requiredColumn : filterClause.getRequiredColumns()) {
+      if (!adapter.isBaseColumn(requiredColumn)) {
+        return JoinFilterAnalysis.createNoPushdownFilterAnalysis(filterClause);
+      }
+    }
+    return new JoinFilterAnalysis(
+        true,
+        false,
+        filterClause,
+        filterClause,
+        null
+    );
+  }
+
+  /**
+   * Potentially rewrite the subfilters of an OR filter so that the whole OR filter can be pushed down to
+   * the base table.
+   *
+   * @param adapter          Adapter for the join
+   * @param orFilter         OrFilter to be rewritten
+   * @param prefixes         Map of table prefixes to clauses
+   * @param equiconditions   Map of equiconditions
+   * @param correlationCache Column correlation analysis cache
+   *
+   * @return A JoinFilterAnalysis indicating how to handle the potentially rewritten filter
+   */
+  @Nullable
```

Comment:
Can remove since this method never returns `null`

Reply:
Removed the unnecessary `@Nullable`